쥬스 메이커 [STEP 1] onetool, 두두

FirstDo opened the pull request:
@1Consumption
안녕하세요 올라프!
onetool, 두두 입니다!
쥬스 메이커 STEP 1을 완료하여 PR 보냅니다!
많이 부족하지만 잘 부탁드려요 ㅎㅎ 😊

## 주요 함수

### struct JuiceMaker

`makeJuice()` : 쥬스를 만드는 함수 
`deleteStockFromRecipeRequired()`: 레시피에 적힌 만큼 재고에서 삭제하는 함수

### class FruitStore

`canMakeJuice()`: 레시피대로 쥬스를 만들수 있는지 재고를 확인하는 함수
`changeFruitStock()`: fruit의 재고를 amount 만큼 낮추는 함수

### ❗️고민한 점

- FruitStore에서 과일을 저장하려고 [Fruit: Int]의 딕셔너리를 사용했는데, 더 좋은 방법이 있을까요? value 값을 매번 바인딩 해서 써주려니 힘듭니다
- API Design Guidelines 에 기반하여 네이밍 을 하려고 고민했습니다.
- FruitStore의 fruits  딕셔너리를 어떤식으로 초기화 해줄지 고민했습니다. 

### ❓궁금한 점

- enum을 쓰면, case가 추가삭제 될때마다, switch문에 case를 계속 추가해줘야해서 관리가 힘들다는 말을 들었습니다.  protocol을 사용해서  더 개선시킬 수 있다는 말을 들었는데, 어떤식으로 하는지 궁금합니다.
    - 원툴은 해당 방식으로 프로젝트를 개선시켜보고 싶어하고, 두두는 굳이 쓸 필요가 있나 라는 생각입니다. 올라프의 생각이 궁금합니다

- 이니셜라이저 사용해서 초기화 하는것과, 프로퍼티의 초기값을 지정하는것의 차이가 있을까요? 

- for문안에 guard문을 사용했는데, 해당 방식이 들여쓰기 2회 위반인가요? 만약 위반이라면, 개선방법이 있을까요?

kimt4580 committed:
📝[DOCS] : Readme.md 작성

Readme.md 작성

kimt4580 committed:
➕[ADD] : Fruit, Juice 열거형 추가

FirstDo committed:
✨[FEAT] : JuiceMaker 구현

kimt4580 committed:
✨[FEAT] : FruitStore 구현

FirstDo committed:
⏪️[RENAME] : 네이밍 수정

kimt4580 committed:
♻️[REFACTOR] : 함수 분리

FirstDo committed:
♻️[REFACTOR] : Fruit에 CaseIterable 추가

1Consumption commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -10,9 +10,7 @@ class ViewController: UIViewController {
-        // Do any additional setup after loading the view.
-    }
-
 
+    }
```
왜 `}`가 삭제되고 추가되었을까요? 커밋에는 공백 하나에도 의미가 있어야합니다.

1Consumption commented on JuiceMaker/JuiceMaker/Model/Fruit.swift:
```diff
@@ -0,0 +1,14 @@
+//  Created by taehun kim on 2022/02/22.
+//
+
+enum Fruit: CaseIterable {
+    case strawberry
+    case kiwi
+    case banana
+    case pineapple
+    case mango
+}
```
열거형을 사용하셨군요~ 사용한 이유를 알 수 있을까요?

kimt4580 replied:
이 프로젝트에서 사용되고 있는 과일이 다섯가지로 정해져있기 때문에, 열거형을 사용하여 관리하는 것이 좋다고 생각했습니다.

1Consumption commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,30 @@
+    var fruits = [Fruit: Int]()
+    
+    init() {
+        let initialStock = 10
```
상수를 활용한 모습 보기 좋습니다 👍

1Consumption commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,30 @@
+                throw JuiceMakerError.outOfStock
+            }
+        }
+    }
```
오... throws를 사용하셨네요. 이건 어떤 역할을 하나요? Bool을 반환하는 것과 차이는 뭔가요?
또한 `조동사 + 동사 원형`은 주로 Bool을 반환하는 메소드나 변수로 사용이 됩니다.

-----------------------------------

Juice라는 개념을 Fruit에서 알아야할까요? FruitStore의 역할은 Fruit에 대한 비즈니스 로직을 담당하는게 아닌가요?

kimt4580 replied:
> 오... throws를 사용하셨네요. 이건 어떤 역할을 하나요? Bool을 반환하는 것과 차이는 뭔가요?

- bool을 반환하게 되면 true와 false만 반환하게 되어 다양한 오류를 핸들링 할 수 없게 됩니다. 하지만 throws를 사용하게 되면 다양한 오류를 핸들링 할 수 있게 됩니다.

> Juice라는 개념을 Fruit에서 알아야할까요? FruitStore의 역할은 Fruit에 대한 비즈니스 로직을 담당하는게 아닌가요?

- 아뇨 몰라도 됩니다. 그래서 canMakeJuice 함수를 없애고, 한 가지 과일의 갯수만 체크하는 checkStock 함수로 변경해주었습니다.

1Consumption commented on JuiceMaker/JuiceMaker/Model/Juice.swift:
```diff
@@ -0,0 +1,35 @@
+    case mangoKiwiJuice
+    case strawberryBananaJuice
+    
+    var recipe: [Fruit: Int] {
+        switch self {
+        case .strawberryJuice:
+            return [.strawberry: 16]
+        case .kiwiJuice:
+            return [.kiwi: 3]
+        case .bananaJuice:
+            return [.banana: 2]
+        case .pineappleJuice:
+            return [.pineapple: 2]
+        case .mangoJuice:
+            return [.mango: 3]
+        case .mangoKiwiJuice:
+            return [.mango: 2, .kiwi: 1]
+        case .strawberryBananaJuice:
+            return [.strawberry: 10, .banana: 1]
+        }
+    }
```
👍👍👍👍 제가 말씀드리려 했던 건데 이미 하셨군요! 이러면 레시피 정보가 한곳으로 모여서 보기 좋네요.

1Consumption commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,30 @@
-
 // 과일 저장소 타입
 class FruitStore {
+    var fruits = [Fruit: Int]()
```
외부에서 알아야하는 값인가요?

kimt4580 replied:
private 으로 접근 제어자를 선언해 주었습니다!

1Consumption commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -2,11 +2,28 @@
 
 // 쥬스 메이커 타입
 struct JuiceMaker {
+    let fruitStore = FruitStore()
```
이부분도 외부에서 알아야하는 값인가요?

kimt4580 replied:
private 으로 접근 제어자를 선언해 주었습니다!

1Consumption commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -2,11 +2,28 @@
+        } catch {
+            print("알 수 없는 오류입니다")
+        }
+    }
```
에러를 catch하는 이유는 뭔가요? 여기서 catch한다면 error에 대한 정보를 알아야하는 사용자에게까지 전달이 될까요?

추가적으로.. 성공 / 실패 여부를 반환하는 `Result`라는 타입도 있습니다.

FirstDo replied:
에러를 viewController에서 catch하도록 변경했습니다

1Consumption commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -2,11 +2,28 @@
+    let fruitStore = FruitStore()
+    
+    func makeJuice(fruitJuice: Juice) {
+        let recipe = fruitJuice.recipe
```
recipe을 넘기기보다는 Fruit를 넘겨서 사용가능한지를 체크하고, 가능하다면 사용하는게 어떨까요?

FirstDo replied:
fruitStore는 과일과 필요량만 받아서 만들수 있는지 없는지 검사하도록 변경했습니다

1Consumption commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -2,11 +2,28 @@
+        for (fruit, amount) in recipe {
+            fruitStore.changeFruitStock(fruit: fruit, amount: amount)
+        }
+    }
```
이 메소드는 밖에서 알아야하나요?

FirstDo replied:
private 으로 바꿔주었습니다!

1Consumption commented on JuiceMaker/JuiceMaker/View/Base.lproj/Main.storyboard:
```diff
@@ -1,9 +1,9 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="19162" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES" useSafeAreas="YES" colorMatched="YES" initialViewController="DCG-dP-Fms">
+<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="19529" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES" useSafeAreas="YES" colorMatched="YES" initialViewController="DCG-dP-Fms">
     <device id="retina6_0" orientation="landscape" appearance="light"/>
     <dependencies>
         <deployment identifier="iOS"/>
-        <plugIn identifier="com.apple.InterfaceBuilder.IBCocoaTouchPlugin" version="19144"/>
+        <plugIn identifier="com.apple.InterfaceBuilder.IBCocoaTouchPlugin" version="19519"/>
```
이 변경사항은 어떤걸 의미하나요? 커밋할 때는 자기가 어떤걸 커밋했는지를 알아야합니다.

FirstDo replied:
앞으로 필요없는건 discard를 하겠습니다!

1Consumption commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,30 @@
+        }
+    }
     
+    func changeFruitStock(fruit: Fruit, amount: Int) {
```
메소드 이름에도 fruit이 들어가있고 매개변수에도 fruit이 들어가있네요. 중복이 발생하지 않게 이름을 바꿀 수 있을까요?

FirstDo replied:
changeStock으로 변경했습니다!

1Consumption commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -2,11 +2,28 @@
+        }
+    }
     
+    func deleteStockFromRecipeRequired(recipe: [Fruit: Int]) {
```
여기도 메소드와 매개변수에 중복되는 이름이 있네요.

FirstDo replied:
deleteStockFrom으로 변경했습니다!

1Consumption reviewed (changes requested):
안녕하세요~ 두두, 원툴! 저도 잘부탁드려요 🙇🙇🙇

올려주신 코드, 커밋 잘 읽었어요.
깃모지가 너무 예쁘네요 🤩

----------------------

먼저 고민한 점, 궁금한 점에 대한 제 생각을 말씀드릴게요!

> FruitStore에서 과일을 저장하려고 [Fruit: Int]의 딕셔너리를 사용했는데, 더 좋은 방법이 있을까요? value 값을 매번 바인딩 해서 써주려니 힘듭니다

Fruit이 유니크하기 때문에 key가 되기 딱 좋은 조건을 가지고 있어서 딕셔너리를 사용하는게 좋아보입니다. 그런데 바인딩을 매번 해준다는게 힘들다는 말은 동의를 못하겠네요.

> enum을 쓰면, case가 추가삭제 될때마다, switch문에 case를 계속 추가해줘야해서 관리가 힘들다는 말을 들었습니다. protocol을 사용해서 더 개선시킬 수 있다는 말을 들었는데, 어떤식으로 하는지 궁금합니다.
원툴은 해당 방식으로 프로젝트를 개선시켜보고 싶어하고, 두두는 굳이 쓸 필요가 있나 라는 생각입니다. 올라프의 생각이 궁금합니다

음... 글쎄요 protocol을 어떻게 사용한다는건지 자세한 정보가 없어서 뭐라 답변드리기 어렵네요. 그런데 enum의 case가 추가 / 삭제 될 때마다 관리를 해줘야함이 불편하다면 protocol도 마찬가지 아닐까요? protocol에 메소드나 변수를 추가해도 해당 protocol을 채택하고있는 객체는 변경사항을 대응해야합니다.

> 이니셜라이저 사용해서 초기화 하는것과, 프로퍼티의 초기값을 지정하는것의 차이가 있을까요?

이니셜라이저는 초기화 하는 로직을 넣을 수 있는 메소드인데, 저장 프로퍼티는 값을 할당하기만 한다는 점이 차이 아닐까요?

> for문안에 guard문을 사용했는데, 해당 방식이 들여쓰기 2회 위반인가요? 만약 위반이라면, 개선방법이 있을까요?

따지자면 들여쓰기 위반이 맞는거 같네요. 사실 for문을 돌면서 throw를 할 이유가 없어보이는 로직이긴 합니다... 이부분은 코멘트로 따로 남겼어요.

--------------------------------

이제 커밋에 대해 아쉬운 점을 말해볼게요.

* [➕[ADD] : Fruit, Juice 열거형 추가](https://github.com/yagom-academy/ios-juice-maker/pull/182/commits/b872f57cdcae55ded4719f2c18d2bac0e578a02c)[](https://github.com/kimt4580)
  * 이 커밋은 한번에 두개의 역할을 하고있는 것처럼 보입니다. `Fruit`와 `Juice`는 꼭 동시에 만들어져야 했나요? 커밋은 최소한의 작업단위여야 합니다.

* [✨[FEAT] : JuiceMaker 구현](https://github.com/FirstDo/ios-juice-maker/commit/88903440774fdf3538c164181a0a695b139b64ec)[](https://github.com/FirstDo)
  * 커밋의 이름에 JuiceMaker의 어떤 부분을 구현했는지에 대한 정보가 없어서 너무 추상적인 네이밍이라고 생각됩니다.
  * 또한 아래와 같은 코드는 커밋하지 않는게 좋아보여요. 메소드 이름도 이름이고, 구현과 무관해보입니다,.
  * <img width="429" alt="image" src="https://user-images.githubusercontent.com/37682858/155180044-923d148f-131b-4c09-be6b-a40db66a276c.png">

* [✨[FEAT] : FruitStore 구현](https://github.com/FirstDo/ios-juice-maker/commit/d852ea51c027ba68aea896c17be65d20321006f6)
  * 이 커밋도 한번에 두개의 역할을 하고 있어요. 그런데 커밋의 이름은 마치 한가지 역할만 하고 있는 것 처럼 보이네요. 다시 한번 말하지만 커밋은 최소한의 작업단위여야 합니다.

* 아래 스샷만 보면 어떤 이름을 수정했는지, 어떤 메소드를 분리했는지 전혀 알수가 없네요. 이런 경우는 커밋 로그를 한번 더 타고 들어가야한다는 불편함이 있습니다.
  * <img width="366" alt="image" src="https://user-images.githubusercontent.com/37682858/155180572-33257d36-382f-4468-8519-520322b5ad99.png"> 
 
-------------------------------------------

나머지 코드에 대한 피드백은 코멘트에 남겼습니다.
고생하셨어요!

kimt4580 committed:
⚡️[IMPROVE] : JuiceMake와 FruitStore에 있는 프로퍼티 fruitStore, fruits를 private으로 변경

FirstDo committed:
✨[FEAT] : FruitStore에 있는 haveEnough 함수 구현

kimt4580 committed:
⚰️ [DEL] :  canMakeJuice 함수 삭제, do-catch문 삭제

FirstDo committed:
♻️[REFACTOR] : JuiceMaker에 canMake 함수 추가

kimt4580 committed:
✏️ [CORRECT] : change name

haveEnough -> checkStock, changeFruitStock -> changeStock, fruits -> fruitStock, makeJuice -> make, deleteStockFromRecipeRequired -> deleteStockFrom

FirstDo committed:
⚰️ [DEL] : canMake 함수 삭제

kimt4580 committed:
⚡️[IMPROVE] : checkStock 함수에서 Bool -> throws 변경

FirstDo committed:
♻️[REFACTOR] : 레시피에 재고가 충분한지 확인하는 부분을 checkEnoughStock으로 분리

kimt4580 committed:
⚡️[IMPROVE] : deleteStockFrom 함수에 private 선언

FirstDo committed:
⚡️[IMPROVE] : checkEnoughStock에 private 접근제어자 추가

1Consumption commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -8,26 +8,26 @@
     }
     
-    func deleteStockFromRecipeRequired(recipe: [Fruit: Int]) {
+    func deleteStockFrom(recipe: [Fruit: Int]) {
```
``` suggestion
func deleteStock(from recipe: [Fruit: Int]) {
```

from을 전달인자로 빼면 더 읽기 좋을거같아요!

FirstDo replied:
말씀대로 변경했습니다!
deleteStock(from recipe: [Fruit: Int])

1Consumption reviewed (approved):
고생하셨습니다~ 늦어져서 죄송해요ㅠㅠ

throws를 적극 활용한 모습 보기 좋습니다. 코드도 훨신 깔끔해졌네요!

코멘트를 하나 남겼는데 이거는 다음 스텝에 반영하도록 하죠.
